index.js

```diff
@@ -145,72 +143,52 @@ ShareDbMongo.prototype._connect = function(mongo, options) {
   // implementing a way to retry
-  var self = this;
+  var connections = [connect(mongo, options.mongoOptions)];
   var mongoPoll = options.mongoPoll;
-  if (mongoPoll) {
-    var tasks = {
-      mongoClient: connect(mongo, options.mongoOptions),
-      mongoPollClient: connect(mongoPoll, options.mongoPollOptions)
+  if (mongoPoll) connections.push(connect(mongoPoll, options.mongoPollOptions));
+
+  return Promise.all(connections).then(function(clients) {
+    var mongoClient = clients[0];
+    var mongoPollClient = clients[1];
+    var result = {
+      mongo: mongoClient,
+      mongoClient: mongoClient,
+      mongoPoll: mongoPollClient,
+      mongoPollClient: mongoPollClient
     };
-    async.parallel(tasks, function(err, results) {
-      if (err) throw err;
-      var mongoClient = results.mongoClient;
-      var mongoPollClient = results.mongoPollClient;
-      if (isLegacyMongoClient(mongoClient)) {
-        self.mongo = self._mongoClient = mongoClient;
-        self.mongoPoll = self._mongoPollClient = mongoPollClient;
-      } else {
-        self.mongo = mongoClient.db();
-        self._mongoClient = mongoClient;
-        self.mongoPoll = mongoPollClient.db();
-        self._mongoPollClient = mongoPollClient;
-      }
-      self._flushPendingConnect();
-    });
-    return;
-  }
-  var finish = function(err, client) {
-    if (err) throw err;
-    if (isLegacyMongoClient(client)) {
-      self.mongo = self._mongoClient = client;
-    } else {
-      self.mongo = client.db();
-      self._mongoClient = client;
+    if (!isLegacyMongoClient(mongoClient)) {
+      result.mongo = mongoClient.db();
+      if (mongoPollClient) result.mongoPoll = mongoPollClient.db();
     }
-    self._flushPendingConnect();
-  };
-  if (typeof mongo === 'function') {
-    mongo(finish);
-    return;
-  }
-  // TODO: Don't pass options directly to mongodb.connect();
-  // only pass options.mongoOptions
-  var mongoOptions = options.mongoOptions || options;
-  connect(mongo, mongoOptions)(finish);
+    return result;
+  });
 };
 
 function connect(mongo, options) {
-  if (typeof mongo === 'function') return mongo;
-  return function(callback) {
-    options = Object.assign({}, options);
-    delete options.mongo;
-    delete options.mongoPoll;
-    delete options.mongoPollOptions;
-    delete options.pollDelay;
-    delete options.disableIndexCreation;
-    delete options.allowAllQueries;
-    delete options.allowJSQueries;
-    delete options.allowAllQueries;
-    delete options.allowAggregateQueries;
-    delete options.getOpsWithoutStrictLinking;
-
-    if (typeof mongodb.connect === 'function') {
-      mongodb.connect(mongo, options, callback);
-    } else {
-      var client = new mongodb.MongoClient(mongo, options);
-      client.connect(callback);
-    }
-  };
+  if (typeof mongo === 'function') {
+    return new Promise(function(resolve, reject) {
+      mongo(function(error, db) {
```

Comment:
The `mongo` function can return either a `MongoClient` or `DB`, so let's name it to reflect that.

Note - It also could be possible for a user's `mongo` function to return a `DB` even if they're on a recent Mongo driver version. It'd be a breaking change if we required it to be a `MongoClient` always.

Reply:
Are you sure about that? The [existing code](https://github.com/share/sharedb-mongo/blob/e72c6bb1a3e9896b29a3f0cbb8f120fe5d53bd63/index.js#L172-L185) looks like it assumes it's always a client (ie it can call `.db()` on the result)